### ISSUE TITLE
vApp error: The requested operation on vApp is not supported in its current state

### ISSUE THREAD

**Issue by pkleanthous** [terraform-providers/terraform-provider-vcd#68]
I'm trying to create a vApp and a VM inside the vApp and it fails as vApp is not ready when terraform tries to execute the next resource.

### Terraform Version
Terraform v0.10.8

### Affected Resource(s)
- vcd_vapp
- vcd_vapp_vm

### Terraform Configuration Files
```hcl
resource "vcd_vapp" "test-demo" {
  name = "test-demo"
}

resource "vcd_vapp_vm" "demo-vm-1" {
  vapp_name     = "${vcd_vapp.test-demo.name}"
  name          = "demo-vm-1"
  catalog_name  = "magic_catalog"
  template_name = "ubuntu-xenial"
  memory        = 2048
  cpus          = 1
  network_name  = "Magic Network"
  depends_on    = ["vcd_vapp.test-demo"]
}
```

### Debug Output
```
2017/11/08 13:19:32 [DEBUG] apply: vcd_vapp.test-demo: executing Apply
2017-11-08T13:19:32.095Z [DEBUG] plugin.terraform-provider-vcd_v1.0.0_x4: 2017/11/08 13:19:32 [DEBUG] Waiting for state to become: [success]
2017/11/08 13:19:33 [TRACE] root: eval: *terraform.EvalWriteState
2017/11/08 13:19:33 [TRACE] root: eval: *terraform.EvalApplyProvisioners
2017/11/08 13:19:33 [TRACE] root: eval: *terraform.EvalIf
2017/11/08 13:19:33 [TRACE] root: eval: *terraform.EvalWriteState
2017/11/08 13:19:33 [TRACE] root: eval: *terraform.EvalWriteDiff
2017/11/08 13:19:33 [TRACE] root: eval: *terraform.EvalApplyPost
2017/11/08 13:19:33 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* vcd_vapp.test-demo: Error Powering Up: &errors.errorString{s:"error powering on vApp: API Error: 400: [ 619fb0ef-167e-4684-9e58-b972cd120ace ] The requested operation on vApp \"test-demo\" is not supported in its current state."}
2017/11/08 13:19:33 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* vcd_vapp.test-demo: Error Powering Up: &errors.errorString{s:"error powering on vApp: API Error: 400: [ 619fb0ef-167e-4684-9e58-b972cd120ace ] The requested operation on vApp \"test-demo\" is not supported in its current state."}
2017/11/08 13:19:33 [TRACE] [walkApply] Exiting eval tree: vcd_vapp.test-demo
```

### Expected Behavior
`vcd_vapp_vm` resource should wait for `vcd_vapp` to succeed before it tries to power it up

### Actual Behavior
`vcd_vapp_vm` resource tries to power up `vcd_vapp` immidiatly

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply`

**Comment by bobbydeveaux** [comment]
Hi @pkleanthous - are you able to let me know what version of VCD you are using? Cheers

**Comment by pkleanthous** [comment]
Hi, @bobbydeveaux thanks for your fast reply. From where can I identify the VCD version?

**Comment by bobbydeveaux** [comment]
Do you have access to the vCloud GUI? If you're administering your VM's, you can press any of the help buttons (blue circle with white q'mark). It opens up the correct help page for the version you're running: e.g. http://pubs.vmware.com/vcd-810/index.jsp?topic=%2Fcom.vmware.vcloud.users.doc_810%2FGUID-B1B05CCA-BE83-4F06-9FFE-789D7E227A76.html

In this case, I can see I'm running 810 (8.1)

**Comment by pkleanthous** [comment]
I confirm that my VCD version is `810`

**Comment by pkleanthous** [comment]
@dn1s Thanks for your reply.

I will try the proposed workaround although if this is a real bug let's keep the issue open until is fixed.

**Comment by opsforge** [comment]
@pkleanthous that's 'normal' for this provider - the calls are issued in short consecution (one for creating a vapp, the other to edit the vapp and add a vm); short enough to cause you to get a 400 as the vapp is still being created when you're trying to edit it. my workaround for this is to use a wrapper around TF that issues a 5 time retry with a 20 second sleep per cycle.

**Comment by Asgoret** [comment]
same issue

**Comment by Asgoret** [comment]
@dn1s and how start them after?

**Comment by dn1s** [comment]
```
resource "vcd_vapp" "test" {
  name = "test"
  power_on = "false"
  count = "${var.node_count}"
}

resource "vcd_vapp_vm" "test" {
  vapp_name = "${element(vcd_vapp.test.*.name, count.index)}"
  name = "${var.node_name}${format("%02d", count.index+1)}.${var.vcd_org}.${var.dns_suffix}"
  template_name = "${var.template_name}"
  catalog_name = "${var.catalog_name}"
  network_name = "net"
  ip = "allocated"
  initscript = "${file("${path.module}/../scripts/initscript")}"
  memory = "${var.node_memory}"
  cpus = "${var.node_cpus}"
  power_on = "true"
  count = "${var.node_count}"
}
```
That's my configuration. So `power_on = false` in `resource "vcd_vapp"` and `power_on = true` in `resource "vcd_vapp_vm"`
This works because if all vms in a vapp are powered on the vapps state changes to powered on as well.

**Comment by Didainius** [comment]
I've seen such issue when environment is quite responsive and even though powering on empty vApp doesn't give a lot (or probably nothing), this bug can probably fixed. I had it reproduced in one environment and will look at solving it.

**Comment by Didainius** [comment]
@Asgoret,
The fix is in our `vmware-stash-while-master-pr-open` branch. If you can build and try out - go ahead.
It should also be soon merged to `master` and will also be included in v2.1 which should happen soon.
